forc-util/src/cli.rs

```diff
@@ -1,38 +1,153 @@
+    } else {
+        let new_build_path = std::process::Command::new("cargo")
+            .env("CLI_TEST", "true")
+            .args(["build", "--message-format=json"])
+            .output()
+            .and_then(|output| {
+                let stdout_str = String::from_utf8_lossy(&output.stdout);
+                let build_messages: Vec<_> = stdout_str
+                    .lines()
+                    .filter_map(|line| serde_json::from_str::<BuildMessage>(line).ok())
+                    .collect();
+
+                let binary_paths: Vec<_> = build_messages
+                    .iter()
+                    .filter(|message| {
+                        message.reason == "compiler-artifact"
+                            && message.target.kind.iter().any(|kind| kind == "bin")
+                    })
+                    .map(|message| message.executable.clone())
+                    .collect();
+
+                binary_paths
+                    .first()
+                    .cloned()
+                    .map(|p| {
+                        std::path::PathBuf::from(p)
+                            .parent()
+                            .unwrap()
+                            .to_str()
+                            .unwrap()
+                            .to_owned()
+                    })
+                    .ok_or_else(|| {
+                        std::io::Error::new(std::io::ErrorKind::Other, "Binary path not found")
+                    })
+            })
+            .unwrap();
+
+        *build_path = Some(new_build_path.clone());
+        format!("{}/{}", new_build_path, bin_name)
+    }
+}
+
 #[macro_export]
-// Let the user format the help and parse it from that string into arguments to create the unit test
+/// Run an arbitrary code block if running in the CLI_TEST environment
+macro_rules! if_cli_test {
+    ($($code:tt)*) => {
+        if option_env!("CLI_TEST").is_some() {
+            $($code)*
+        }
+    };
+}
+
+#[macro_export]
+/// Let the user format the help and parse it from that string into arguments to create the unit
+/// test.
+///
+/// The list of examples is a list of tuples where the first element is the description of the test
+/// (in plain English) followed by the command to be executed and the arguments to be passed to the
+/// command. Optionally, the expected output of the command can be passed as well. This examples are
+/// part of the help message of the CLI.
+///
+/// Each example is also converted into a unit test. The test invokes the CLI command externall
+/// (there is no `#[cfg(test)]` since the command is an external process and unaware of the test
+/// context). The `option_env!("CLI_TEST").is_some()` expression can be used to detect if the command is
+/// being executed from the CLI_TEST environment and take a different path (for instance to mock a
+/// user given input response).
+///
+/// This macro also takes a list of examples and a setup and teardown block. The setup code block is
```

Comment:
I remember you saying you were thinking of removing the teardown block. Is this still the case? If so we should remove that from the doc comment.